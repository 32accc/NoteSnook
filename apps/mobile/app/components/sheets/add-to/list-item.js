/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import React, { useState } from "react";
import { View } from "react-native";
import { useThemeStore } from "../../../stores/use-theme-store";
import { SIZE } from "../../../utils/size";
import { IconButton } from "../../ui/icon-button";
import { PressableButton } from "../../ui/pressable";
import Heading from "../../ui/typography/heading";
import Paragraph from "../../ui/typography/paragraph";
import { useSelectionContext } from "./context";
import { FilteredList } from "./filtered-list";
import { useItemSelectionStore } from "./store";

const SelectionIndicator = ({ item, hasNotes, selectItem, onPress }) => {
  const itemState = useItemSelectionStore((state) => state.itemState[item.id]);
  const multiSelect = useItemSelectionStore((state) => state.multiSelect);

  const isSelected = itemState === "selected";
  const isIntermediate = itemState === "intermediate";
  const isRemoved = !isSelected && hasNotes;
  const colors = useThemeStore((state) => state.colors);

  return (
    <IconButton
      size={22}
      customStyle={{
        marginRight: 5,
        width: 23,
        height: 23
      }}
      color={
        isRemoved
          ? colors.red
          : isIntermediate || isSelected
          ? colors.accent
          : colors.icon
      }
      onPress={() => {
        if (multiSelect) return selectItem();
        onPress?.(item);
      }}
      testID={
        isRemoved
          ? "close-circle-outline"
          : isSelected
          ? "check-circle-outline"
          : isIntermediate
          ? "minus-circle-outline"
          : "checkbox-blank-circle-outline"
      }
      name={
        isRemoved
          ? "close-circle-outline"
          : isSelected
          ? "check-circle-outline"
          : isIntermediate
          ? "minus-circle-outline"
          : "checkbox-blank-circle-outline"
      }
    />
  );
};

export const ListItem = ({
  item,
  index,
  icon,
  infoText,
  hasSubList,
  onPress,
  onScrollEnd,
  getListItems,
  style,
  type,
  sublistItemType,
  onAddItem,
  getSublistItemProps,
  hasHeaderSearch,
  onAddSublistItem,
  hasNotes,
  sheetRef
}) => {
  const { toggleSelection } = useSelectionContext();
  const multiSelect = useItemSelectionStore((state) => state.multiSelect);

  const colors = useThemeStore((state) => state.colors);
  const [expanded, setExpanded] = useState(false);

  function selectItem() {
    toggleSelection(item);
  }

  return (
    <View
      style={{
        overflow: "hidden",
        marginBottom: 10,
        ...style
      }}
    >
      <PressableButton
        onPress={() => {
          if (hasSubList) return setExpanded(!expanded);
          if (multiSelect) return selectItem();
          onPress?.(item);
        }}
        type={type}
        onLongPress={() => {
          useItemSelectionStore.getState().setMultiSelect(true);
          selectItem();
        }}
        customStyle={{
          height: style?.height || 50,
          width: "100%",
          alignItems: "flex-start"
        }}
      >
        <View
          style={{
            width: "100%",
            height: 50,
            justifyContent: "space-between",
            flexDirection: "row",
            alignItems: "center",
            paddingHorizontal: 12
          }}
        >
          <View
            style={{
              flexDirection: "row",
              alignItems: "center"
            }}
          >
            <SelectionIndicator
              hasNotes={hasNotes}
              onPress={onPress}
              item={item}
              selectItem={selectItem}
            />
            <View>
              {hasSubList && expanded ? (
                <Heading size={SIZE.md}>{item.title}</Heading>
              ) : (
                <Paragraph size={SIZE.sm}>{item.title}</Paragraph>
              )}

              {infoText ? (
                <Paragraph size={SIZE.xs} color={colors.icon}>
                  {infoText}
                </Paragraph>
              ) : null}
            </View>
          </View>

          <View
            style={{
              flexDirection: "row"
            }}
          >
            {onAddSublistItem ? (
              <IconButton
                name={"plus"}
                testID="add-item-icon"
                color={colors}
                size={SIZE.xl}
                onPress={() => {
                  onAddSublistItem(item);
                }}
              />
            ) : null}
            {icon ? (
              <IconButton
                name={icon(expanded).name}
                color={icon(expanded).color}
                size={icon(expanded).size || SIZE.xl}
                onPress={
                  hasSubList
                    ? () => setExpanded(!expanded)
                    : icon(expanded).onPress
                }
              />
            ) : null}
          </View>
        </View>
      </PressableButton>

      {expanded && hasSubList ? (
        <FilteredList
          nestedScrollEnabled
          data={getListItems(item)}
          keyboardShouldPersistTaps="always"
          keyboardDismissMode="none"
          onMomentumScrollEnd={onScrollEnd}
          style={{
            width: "95%",
            alignSelf: "flex-end",
            maxHeight: 250
          }}
          itemType={sublistItemType}
          hasHeaderSearch={hasHeaderSearch}
          renderItem={({ item, index }) => (
            <ListItem
              item={item}
              {...getSublistItemProps(item)}
              index={index}
              onScrollEnd={onScrollEnd}
            />
          )}
          onAddItem={onAddItem}
        />
      ) : null}
    </View>
  );
};
