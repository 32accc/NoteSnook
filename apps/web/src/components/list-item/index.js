/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { Box, Flex, Text } from "@theme-ui/components";
import * as Icon from "../icons";
import {
  store as selectionStore,
  useStore as useSelectionStore
} from "../../stores/selection-store";
import { useMenuTrigger } from "../../hooks/use-menu";
import Config from "../../utils/config";
import { db } from "../../common/db";
import * as clipboard from "clipboard-polyfill/text";
import { useRef } from "react";

function debugMenuItems(type) {
  if (!type) return [];
  return [
    {
      key: "copy-data",
      title: () => "Copy data",
      icon: Icon.Copy,
      onClick: async ({ [type]: item }) => {
        if (type === "note" && item.contentId) {
          item.additionalData = {
            content: db.debug.strip(await db.content.raw(item.contentId))
          };
        }
        item.additionalData = {
          ...item.additionalData,
          lastSynced: await db.lastSynced()
        };
        await clipboard.writeText(db.debug.strip(item));
      }
    }
  ];
}

function ListItem(props) {
  const {
    colors: { text, background, primary } = {
      primary: "primary",
      text: "text",
      background: "background"
    },
    isFocused,
    isCompact,
    isDisabled,
    isSimple
  } = props;

  const listItemRef = useRef();
  const { openMenu, target } = useMenuTrigger();
  const isMenuTarget = target && target === listItemRef.current;

  const isSelected = useSelectionStore((store) => {
    const inInSelection =
      store.selectedItems.findIndex((item) => props.item.id === item.id) > -1;
    return isFocused
      ? store.selectedItems.length > 1 && inInSelection
      : inInSelection;
  });

  return (
    <Flex
      id={`id_${props.item.id}`}
      className={isSelected ? "selected" : ""}
      ref={listItemRef}
      onContextMenu={(e) => {
        e.preventDefault();
        e.stopPropagation();

        let items = props.menu?.items?.slice() || [];
        let title = undefined;
        let selectedItems = selectionStore
          .get()
          .selectedItems.filter((i) => i.type === props.item.type);

        if (selectedItems.findIndex((i) => i.id === props.item.id) === -1) {
          selectedItems = [];
          selectedItems.push(props.item);
        }

        if (selectedItems.length > 1) {
          title = `${selectedItems.length} items selected`;
          items = items.filter((item) => item.multiSelect);
        } else if (Config.get("debugMode", false)) {
          items.push(...debugMenuItems(props.item.type));
        }

        if (items.length <= 0) return;

        openMenu(items, {
          title,
          items: selectedItems,
          target: listItemRef.current,
          ...props.menu?.extraData
        });
      }}
      pl={1}
      pr={2}
      py={1}
      mb={isCompact ? 0 : 1}
      tabIndex={-1}
      sx={{
        height: "inherit",
        cursor: "pointer",
        position: "relative",
        overflow: "hidden",
        maxWidth: "100%",

        flexDirection: isCompact ? "row" : "column",
        justifyContent: isCompact ? "space-between" : "center",
        alignItems: isCompact ? "center" : undefined,

        opacity: isDisabled ? 0.7 : 1,

        borderLeft: "5px solid",
        borderLeftColor: isFocused ? primary : "transparent",
        ml: "2px",
        mr: "1px",

        backgroundColor: isSelected
          ? "shade"
          : isMenuTarget || isFocused
          ? "hover"
          : background,

        ":hover": {
          backgroundColor: isSelected ? "textSelection" : "hover"
        },
        ":focus": {
          backgroundColor: isSelected ? "textSelection" : "hover"
        },
        ":focus-visible": {
          outline: `1px solid var(--${
            primary === "primary" ? "dimPrimary" : primary
          })`,
          backgroundColor: isSelected ? "textSelection" : background
        }
      }}
      onKeyPress={(e) => {
        if (e.key !== "Enter") {
          if (props.onKeyPress) props.onKeyPress(e);
        }
      }}
      onClick={(e) => {
        if (!e.metaKey && !e.shiftKey && !e.ctrlKey && props.onClick) {
          props.onClick();
        }
      }}
      data-test-id={`list-item`}
    >
      {!isCompact && props.header}

      <Text
        data-test-id={`title`}
        variant={isSimple || isCompact ? "body" : "subtitle"}
        sx={{
          whiteSpace: "nowrap",
          overflow: "hidden",
          textOverflow: "ellipsis",
          fontWeight: isCompact || isSimple ? "body" : "bold",
          color: text,
          display: "block"
        }}
      >
        {props.title}
      </Text>

      {!isSimple && !isCompact && props.body && (
        <Text
          as="p"
          variant="body"
          data-test-id={`description`}
          sx={{
            lineHeight: `1.2rem`,
            overflow: "hidden",
            textOverflow: "ellipsis",
            whiteSpace: "pre-wrap",
            position: "relative",
            display: "-webkit-box",
            WebkitLineClamp: 4,
            WebkitBoxOrient: "vertical"
          }}
        >
          {props.body}
        </Text>
      )}
      {props.footer ? (
        <Box
          ml={isCompact ? 1 : 0}
          mt={isCompact ? 0 : 1}
          sx={{ flexShrink: 0 }}
        >
          {props.footer}
        </Box>
      ) : null}
    </Flex>
  );
}
export default ListItem;
