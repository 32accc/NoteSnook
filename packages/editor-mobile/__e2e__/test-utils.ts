/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import {
  AndroidDevice,
  BrowserContext,
  TestInfo,
  _android as android,
  test
} from "@playwright/test";
import { Keyboard, SUPPORTED_KEYBOARDS, getKeyboards } from "./keyboard-utils";
import * as ip from "ip";
import { DeviceSize, getDeviceSize } from "./device-utils";

type AndroidTestArgs = {
  chrome: BrowserContext;
  device: AndroidDevice;
  baseURL: string;
  keyboards: Keyboard[];
  size: DeviceSize;
};

const SERVER_ADDRESS = `http://${ip.address("Wi-Fi", "ipv4")}:3000`;

export function androidTest(
  title: string,
  testFunction: (
    args: AndroidTestArgs,
    testInfo: TestInfo
  ) => Promise<void> | void
) {
  test(title, async ({ channel: _ }, testInfo) => {
    const [device] = await android.devices();
    if (!device) {
      console.error("Please connect an Android device or emulator.");
      return;
    }

    const keyboards = await getKeyboards(device);
    if (!keyboards.length) {
      console.error(
        "No supported keyboard found. Please install one of",
        SUPPORTED_KEYBOARDS
      );
      return;
    }

    const deviceSize = await getDeviceSize(device);

    await device.shell("am force-stop com.android.chrome");
    const chrome = await device.launchBrowser({});

    await testFunction(
      {
        chrome: chrome,
        device,
        keyboards,
        baseURL: SERVER_ADDRESS,
        size: deviceSize
      },
      testInfo
    );

    await device.shell("am force-stop com.android.chrome");
    await device.close();
  });
}
