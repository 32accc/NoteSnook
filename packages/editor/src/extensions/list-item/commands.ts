/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { EditorState } from "prosemirror-state";
import { NodeType } from "prosemirror-model";
import { Editor } from "@tiptap/core";
import {
  findParentNodeOfType,
  hasParentNodeOfType
} from "../../utils/prosemirror";

// WORKAROUND: if we're at the start of a list item, we need to either
// backspace directly to an empty list item above, or outdent this node
export function onBackspacePressed(
  editor: Editor,
  name: string,
  type: NodeType
) {
  const { selection } = editor.state;
  const { empty, $from } = selection;

  if (
    !empty ||
    !isInside(name, type, editor.state) ||
    $from.parentOffset !== 0 ||
    !isFirstChildOfParent(editor.state) ||
    !editor.can().liftListItem(type)
  )
    return false;

  const isEmpty = isListItemEmpty(type, editor.state);

  if (isEmpty) {
    if (isFirstOfType(type, editor.state)) {
      const parentList = getListFromListItem(type, editor.state);
      if (!parentList) return false;
      if (parentList.childCount > 1) {
        return editor.commands.liftListItem(type);
      }
    }

    return editor.chain().joinBackward().joinBackward().run();
  } else if (isFirstOfType(type, editor.state)) {
    return editor.commands.liftListItem(type);
  } else {
    const block = findParentNodeOfType(type)(selection);
    if (block && block.start === $from.pos - 1) {
      // we have to run join backward twice because on the first join
      // the two list items are joined i.e., the editor just puts their
      // paragraphs next to each other. The next join merges the paragraphs
      // like it should be.
      return editor.chain().joinBackward().joinBackward().run();
    }
  }
  return false;
}

export function onArrowUpPressed(editor: Editor, name: string, type: NodeType) {
  const { selection } = editor.state;
  const { empty } = selection;

  if (!empty || !isFirstOfType(type, editor.state)) return false;
  const parentList = getListFromListItem(type, editor.state);
  if (editor.state.doc.firstChild === parentList)
    return editor.commands.insertContentAt(0, "<p></p>");
  return false;
}

function isInside(name: string, type: NodeType, state: EditorState) {
  const { $from } = state.selection;

  const node = type || state.schema.nodes[name];
  const { paragraph } = state.schema.nodes;

  return (
    hasParentNodeOfType(node)(state.selection) &&
    $from.parent.type === paragraph
  );
}

function isFirstChildOfParent(state: EditorState): boolean {
  const { $from } = state.selection;
  return $from.depth > 1
    ? $from.parentOffset === 0 || $from.index($from.depth - 1) === 0
    : true;
}

const isFirstOfType = (type: NodeType, state: EditorState) => {
  const block = findParentNodeOfType(type)(state.selection);
  if (!block) return false;
  const { pos } = block;
  const resolved = state.doc.resolve(pos);
  return !resolved.nodeBefore;
};

const getListFromListItem = (type: NodeType, state: EditorState) => {
  const block = findParentNodeOfType(type)(state.selection);
  if (!block) return undefined;
  const { pos } = block;
  const resolved = state.doc.resolve(pos);
  if (
    !resolved.parent.type.spec.group ||
    resolved.parent.type.spec.group?.indexOf("list") <= -1
  )
    return undefined;

  return resolved.parent;
};

function isListItemEmpty(type: NodeType, state: EditorState) {
  const block = findParentNodeOfType(type)(state.selection);
  if (!block) return false;
  const { node } = block;
  return !node.textContent.length;
}
