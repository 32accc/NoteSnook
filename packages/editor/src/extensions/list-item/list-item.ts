/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { ListItem as TiptapListItem } from "@tiptap/extension-list-item";

export const ListItem = TiptapListItem.extend({
  addKeyboardShortcuts() {
    return {
      ...this.parent?.(),
      Tab: (props) => {
        const { editor } = props;
        const { state } = editor;
        const { selection } = state;
        const { $from } = selection;

        if ($from.parent.type.name === "codeblock") return false;

        return this.parent?.()?.Tab(props) || false;
      }
    };
  }
});
