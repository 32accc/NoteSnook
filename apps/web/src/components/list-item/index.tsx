/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { Box, Flex, Text } from "@theme-ui/components";
import {
  store as selectionStore,
  useStore as useSelectionStore
} from "../../stores/selection-store";
import { useMenuTrigger } from "../../hooks/use-menu";
import React, { useRef } from "react";
import { SchemeColors } from "@notesnook/theme";
import { MenuItem } from "@notesnook/ui";
import { alpha } from "@theme-ui/color";
import { Item } from "@notesnook/core/dist/types";

type ListItemProps = {
  colors?: {
    heading: SchemeColors;
    accent: SchemeColors;
    background: SchemeColors;
  };
  isFocused?: boolean;
  isCompact?: boolean;
  isDisabled?: boolean;
  isSimple?: boolean;
  item: Item;

  onKeyPress?: (e: React.KeyboardEvent<HTMLDivElement>) => void;
  onClick?: () => void;
  title: string | JSX.Element;
  header?: JSX.Element;
  body?: JSX.Element | string;
  footer?: JSX.Element;

  menuItems?: (item: any, items?: any[]) => MenuItem[];
};

function ListItem(props: ListItemProps) {
  const {
    colors: { heading, background, accent } = {
      heading: "heading",
      accent: "accent",
      background: "background"
    },
    isFocused,
    isCompact,
    isDisabled,
    isSimple,
    item
  } = props;

  const listItemRef = useRef<HTMLDivElement>(null);
  const { openMenu, target } = useMenuTrigger();
  const isMenuTarget = target && target === listItemRef.current;

  const isSelected = useSelectionStore((store) => {
    const isInSelection =
      store.selectedItems.findIndex((item) => item.id === props.item.id) > -1;
    return isFocused
      ? store.selectedItems.length > 1 && isInSelection
      : isInSelection;
  });
  const selected = isSelected || isMenuTarget || isFocused;

  return (
    <Flex
      id={`id_${item.id}`}
      className={isSelected ? "selected" : ""}
      ref={listItemRef}
      onContextMenu={(e) => {
        e.preventDefault();
        e.stopPropagation();

        let title = undefined;
        let selectedItems = selectionStore
          .get()
          .selectedItems.filter((i) => i.type === item.type);

        if (selectedItems.findIndex((i) => i.id === item.id) === -1) {
          selectedItems = [];
          selectedItems.push(item);
        }

        let menuItems = props.menuItems?.(item, selectedItems);

        if (selectedItems.length > 1) {
          title = `${selectedItems.length} items selected`;
          menuItems = menuItems?.filter((i) => i.multiSelect === true);
        }

        if (!menuItems) return;

        openMenu(menuItems, {
          title
        });
      }}
      pl={1}
      pr={2}
      py={1}
      mb={isCompact ? 0 : 0}
      tabIndex={-1}
      dir="auto"
      sx={{
        height: "inherit",
        cursor: "pointer",
        position: "relative",
        overflow: "hidden",
        maxWidth: "100%",

        flexDirection: isCompact ? "row" : "column",
        justifyContent: isCompact ? "space-between" : "center",
        alignItems: isCompact ? "center" : undefined,

        opacity: isDisabled ? 0.7 : 1,

        borderLeft: "5px solid",
        borderLeftColor: isFocused ? accent : "transparent",

        backgroundColor: selected ? "background-selected" : background,

        ":hover": {
          backgroundColor: selected ? "hover-selected" : "hover"
        },
        ":focus": {
          backgroundColor: selected ? "hover-selected" : "hover"
        },
        ":focus-visible": {
          outline: `1px solid`,
          outlineColor: accent === "accent" ? "accent" : alpha("accent", 0.7),
          backgroundColor:
            isSelected || isFocused ? "background-selected" : background
        }
      }}
      onKeyPress={(e) => {
        if (e.key !== "Enter") {
          if (props.onKeyPress) props.onKeyPress(e);
        }
      }}
      onClick={(e) => {
        if (!e.metaKey && !e.shiftKey && !e.ctrlKey && props.onClick) {
          props.onClick();
        }
      }}
      data-test-id={`list-item`}
    >
      {!isCompact && props.header}

      <Text
        data-test-id={`title`}
        variant={isSimple || isCompact ? "body" : "subtitle"}
        sx={{
          whiteSpace: "nowrap",
          overflow: "hidden",
          textOverflow: "ellipsis",
          fontWeight: isCompact || isSimple ? "body" : "bold",
          color:
            selected && heading === "heading" ? `${heading}-selected` : heading,
          display: "block"
        }}
      >
        {props.title}
      </Text>

      {!isSimple && !isCompact && props.body && (
        <Text
          as="p"
          variant="body"
          data-test-id={`description`}
          sx={{
            color: selected ? "paragraph-selected" : "paragraph",
            lineHeight: `1.2rem`,
            overflow: "hidden",
            textOverflow: "ellipsis",
            whiteSpace: "pre-wrap",
            position: "relative",
            display: "-webkit-box",
            WebkitLineClamp: 4,
            WebkitBoxOrient: "vertical"
          }}
        >
          {props.body}
        </Text>
      )}
      {props.footer ? (
        <Box
          ml={isCompact ? 1 : 0}
          mt={isCompact ? 0 : 1}
          sx={{ flexShrink: 0 }}
        >
          {props.footer}
        </Box>
      ) : null}
    </Flex>
  );
}
export default ListItem;
