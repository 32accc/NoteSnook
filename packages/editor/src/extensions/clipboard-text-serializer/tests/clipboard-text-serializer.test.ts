/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { test } from "vitest";
import { createEditor, h } from "../../../../test-utils";
import OrderedList from "../../ordered-list";
import { ListItem } from "../../list-item";
import { getTextBetween, transformCopied, DOMSerializer2 } from "../index";
import { Paragraph } from "../../paragraph";

test("copied list items shouldn't contain extra newlines", (t) => {
  const { editor } = createEditor({
    initialContent: h("div", [
      h("ol", [
        h("li", [
          h("p", ["This is line: ", h("em", ["number 1."])]),
          h("p", ["And this is line number 2."])
        ]),
        h("li", [h("p", ["This is line number 3."])])
      ])
    ]).innerHTML,
    extensions: {
      orderedList: OrderedList,
      listItem: ListItem
    }
  });

  const serializer = DOMSerializer2.fromSchema(editor.view.state.schema);
  t.expect(
    new XMLSerializer().serializeToString(
      serializer.serializeFragment(
        editor.state.doc.slice(0, editor.state.doc.nodeSize - 2).content
      )
    )
  ).toBe(
    "<ol><li><p>This is line: <em>number 1.</em></p><p>And this is line number 2.</p></li><li>This is line number 3.</li></ol>"
  );

  t.expect(
    getTextBetween(
      editor.state.doc.slice(0, editor.state.doc.nodeSize - 2),
      editor.schema
    )
  ).toBe(`This is line: number 1.
And this is line number 2.
This is line number 3.`);
});

test("copying a single list item shouldn't copy the list metadata", (t) => {
  const { editor } = createEditor({
    initialContent: h("div", [h("ol", [h("li", ["Hello"])])]).innerHTML,
    extensions: {
      orderedList: OrderedList,
      listItem: ListItem
    }
  });

  t.expect(
    transformCopied(
      editor.state.doc.slice(0, editor.state.doc.nodeSize - 2)
    ).toJSON()
  ).toMatchSnapshot();
});

const createParagraphs = (spacing: "double" | "single") => [
  h("p", ["I am paragraph 1."], { "data-spacing": spacing }),
  h("p", ["I am paragraph 2."], { "data-spacing": spacing }),
  h("p", ["I am paragraph 3."], { "data-spacing": spacing })
];

const createParagraphsWithSpaces = (spacing: "double" | "single") => [
  h("p", ["I am paragraph 1."], { "data-spacing": spacing }),
  h("p", [h("br")], { "data-spacing": spacing }),
  h("p", [h("br")], { "data-spacing": spacing }),
  h("p", ["I am paragraph 2."], { "data-spacing": spacing }),
  h("p", [h("br")], { "data-spacing": spacing }),
  h("p", [h("br")], { "data-spacing": spacing }),
  h("p", ["I am paragraph 3."], { "data-spacing": spacing })
];

const paragraphTestCases = [
  {
    spacing: "double",
    content: createParagraphs("double"),
    expectedHtml: `<p data-spacing="double">I am paragraph 1.</p><p data-spacing="double">I am paragraph 2.</p><p data-spacing="double">I am paragraph 3.</p>`,
    expectedText: `I am paragraph 1.\n\nI am paragraph 2.\n\nI am paragraph 3.`
  },
  {
    spacing: "single",
    content: createParagraphs("single"),
    expectedHtml: `<p data-spacing="single">I am paragraph 1.<br>I am paragraph 2.<br>I am paragraph 3.</p>`,
    expectedText: `I am paragraph 1.\nI am paragraph 2.\nI am paragraph 3.`
  },
  {
    spacing: "double",
    content: createParagraphsWithSpaces("double"),
    expectedHtml: `<p data-spacing="double">I am paragraph 1.</p><p data-spacing="double"></p><p data-spacing="double"></p><p data-spacing="double">I am paragraph 2.</p><p data-spacing="double"></p><p data-spacing="double"></p><p data-spacing="double">I am paragraph 3.</p>`,
    expectedText: `I am paragraph 1.\n\n\n\nI am paragraph 2.\n\n\n\nI am paragraph 3.`
  },
  {
    spacing: "single",
    content: createParagraphsWithSpaces("single"),
    expectedHtml: `<p data-spacing="single">I am paragraph 1.<br><br><br>I am paragraph 2.<br><br><br>I am paragraph 3.</p>`,
    expectedText: `I am paragraph 1.\n\n\nI am paragraph 2.\n\n\nI am paragraph 3.`
  },
  {
    spacing: "mixed",
    content: [
      ...createParagraphs("double"),
      ...createParagraphs("single"),
      ...createParagraphs("double")
    ],
    expectedHtml: `<p data-spacing="double">I am paragraph 1.</p><p data-spacing="double">I am paragraph 2.</p><p data-spacing="double">I am paragraph 3.<br>I am paragraph 1.<br>I am paragraph 2.<br>I am paragraph 3.</p><p data-spacing="double">I am paragraph 1.</p><p data-spacing="double">I am paragraph 2.</p><p data-spacing="double">I am paragraph 3.</p>`,
    expectedText: `I am paragraph 1.\n\nI am paragraph 2.\n\nI am paragraph 3.\nI am paragraph 1.\nI am paragraph 2.\nI am paragraph 3.\n\nI am paragraph 1.\n\nI am paragraph 2.\n\nI am paragraph 3.`
  },
  {
    spacing: "mixed 2",
    content: [
      h("p", ["I am paragraph 1."], { "data-spacing": "double" }),
      h("p", ["I am paragraph 2."], { "data-spacing": "single" }),
      h("p", ["I am paragraph 3."], { "data-spacing": "double" }),
      h("p", ["I am paragraph 4."], { "data-spacing": "single" }),
      h("p", ["I am paragraph 5."], { "data-spacing": "single" }),
      h("p", ["I am paragraph 6."], { "data-spacing": "double" })
    ],
    expectedHtml: `<p data-spacing="double">I am paragraph 1.<br>I am paragraph 2.</p><p data-spacing="double">I am paragraph 3.<br>I am paragraph 4.<br>I am paragraph 5.</p><p data-spacing="double">I am paragraph 6.</p>`,
    expectedText: `I am paragraph 1.\nI am paragraph 2.\n\nI am paragraph 3.\nI am paragraph 4.\nI am paragraph 5.\n\nI am paragraph 6.`
  }
];
for (const testCase of paragraphTestCases) {
  test(`copying should respect paragraph spacing (${testCase.spacing})`, (t) => {
    const { editor } = createEditor({
      initialContent: h("div", testCase.content).innerHTML,
      extensions: {
        paragraph: Paragraph,
        hardBreak: false
      }
    });

    const serializer = DOMSerializer2.fromSchema(editor.view.state.schema);
    t.expect(
      new XMLSerializer().serializeToString(
        serializer.serializeFragment(
          editor.state.doc.slice(0, editor.state.doc.nodeSize - 2).content
        )
      )
    ).toBe(testCase.expectedHtml);

    t.expect(
      getTextBetween(
        editor.state.doc.slice(0, editor.state.doc.nodeSize - 2),
        editor.schema
      )
    ).toBe(testCase.expectedText);
  });
}
