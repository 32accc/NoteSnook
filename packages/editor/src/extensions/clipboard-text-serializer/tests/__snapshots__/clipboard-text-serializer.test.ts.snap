// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`copying a single list item shouldn't copy the list metadata 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "Hello",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
}
`;
